### PR TITLE
Added hangouts message sender

### PR DESCRIPTION
**Pull request by Sagox**
- Added abstract MessageSender class
- Added HangoutsMessageSender class

### CODE REVIEW

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+          .setOnClick(onClick);
+      final Button widget = new Button().setTextButton(button);
+      widgets.add(new WidgetMarkup().setButtons(Collections.singletonList((widget))));
+      final Section section = new Section().setWidgets(widgets);
+      sectionList.add(section);
+    }
+    final Card card =
+        (new Card()).setSections(Collections.unmodifiableList(sectionList));
+    final Message message = new Message().setCards(Collections.singletonList(card));
+    // the description is sent first followed by the options
```
Comment (yogicxl) [inline services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java]: create a private foo for generating card message and test that.

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// class to handle sending of messages to hangouts users
+
+@Component
+public class HangoutsMessageSender extends MessageSender{
+  private static String CHAT_SCOPE;
+  private static GoogleCredentials credentials;
+  private static HttpRequestInitializer requestInitializer;
+  private static HangoutsChat chatService;
+
+  HangoutsMessageSender(@Value("${hangoutsAPIScope}") final String apiScope,
```
Comment (yogicxl) [inline services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java]: ChatService should be injected. Then you could mock it and test.

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    requestInitializer = new HttpCredentialsAdapter(credentials);
+    chatService = new HangoutsChat.Builder(
+        GoogleNetHttpTransport.newTrustedTransport(),
+        JacksonFactory.getDefaultInstance(), requestInitializer)
+        .setApplicationName("chatbot").build();
+  }
+
+  // send message to a user using their chatClientGeneratedID (spaceID for hangouts)
+  @Override
+  public void sendMessage(final String chatClientGeneratedID, final String msg) throws IOException {
```
Comment (yogicxl) [inline services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java:39]: Overload sendMessage:
sendMessage(id, str);
sendMessage(id,Message);

Then sendMessage(id,str) can call sendMessage.

File: services/src/main/java/com/chatbot/services/MessageSenders/MessageSender.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.chatbot.services.MessageSenders;
+
+import java.io.IOException;
+
+abstract class MessageSender {
+    abstract void sendMessage(final String chatClientGeneratedID, final String msg)
```
Comment (yogicxl) [inline services/src/main/java/com/chatbot/services/MessageSenders/MessageSender.java:6]: I feel you could rename sendMessageBasedOnMessageType. could have three args: ID, msg, msg_type. msg_type could be an enum {TEXT=1,CARD=2,CARD_WITH_TEXT_HEADER=3}.

Impl could be like:
sendMessageBasedOnMessageType () {
List<Message> m = GenerateMessageListByType(string msg, msgtype);
m.stream().map(m -> SendMessage(id,m));
}
GenerateMessageListByType(string msg, msgtype) {
List<Message> m;
switch (msgtype) {
case TEXT:
m.add(generateTextMessge()); break;
case CARD:
m.add(generateCardMessage());break;
case CARD_WITH_HEADER:
m.add(generateTextMessge());
m.add(generateCardMessage()); break;
}
return m;
}


}
}

**Review by yogicxl — CHANGES_REQUESTED** [review]
Please make the mentioned changes.

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java (hunk 1 above)
Comment (Sagox) [reply services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java]: updated

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java (hunk 2 above)
Comment (Sagox) [reply services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java]: is now loaded as a component on application start up

File: services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java (hunk 3 above)
Comment (Sagox) [reply services/src/main/java/com/chatbot/services/MessageSenders/HangoutsMessageSender.java:39]: updated

File: services/src/main/java/com/chatbot/services/MessageSenders/MessageSender.java (hunk 4 above)
Comment (Sagox) [reply services/src/main/java/com/chatbot/services/MessageSenders/MessageSender.java:6]: updated

**Review by yogicxl — APPROVED** [review]

**Pull request merged**